### PR TITLE
Add message body check (400) for http calls

### LINKED ISSUE #198: improve error handling for /v1/chain/get_block_info
There should be no 500 errors due to bad user input.  See previous discussion on error handling for /v1/chain/get_transaction_status 

```
curl -sd '{x"block_num": "42"}' https://jungle4.api.eosnation.io/v1/chain/get_block_info | jq
{
  "code": 500,
  "message": "Internal Service Error",
  "error": {
    "code": 4,
    "name": "parse_error_exception",
    "what": "Parse Error",
    "details": [
      {
        "message": "Expected '\"' but read 'x'",
        "file": "json.cpp",
        "line_number": 97,
        "method": "stringFromStream"

```

Could not find block 0?

```
$ curl -sd '{}' https://jungle4.api.eosnation.io/v1/chain/get_block_info | jq
{
  "code": 400,
  "message": "Unknown Block",
  "error": {
    "code": 3100002,
    "name": "unknown_block_exception",
    "what": "Unknown block",
    "details": [
      {
        "message": "Could not find block: 0",
        "file": "chain_plugin.cpp",
        "line_number": 2084,
        "method": "get_block_info"
      }
    ]
  }
}
```

### LINKED ISSUE #199: Backport Adding message body check (400) for http calls
https://github.com/EOSIO/eos/pull/9033

For Issue: https://github.com/EOSIO/eos/issues/8970
Add parameters check for all http requests.

Changes:
apply parameters check for every http calls.
a) if empty body is not allowed, then throw exception for empty body
b) if empty body is required, then set empty body as "{}" for string and default for other types.
c) if empty body is allowed, then returns same as empty body is required for empty body, and returns same as empty is not allowed for non-empty body.

### PR DESCRIPTION
**Pull request by heifner**
Apply parameters check for every http calls.
* If empty body is not allowed, then report `"A Request body is required"` with error code `400`.
* Require JSON body to contain data for JSON object, `{}` will report: `"A Request body is required"` with error code `400`.
* If unable to parse JSON object for http param report: `"Unable to parse valid input from POST body"` with error code `400`. 

- Resolves #198 
- Resolves #199 
- Backports https://github.com/EOSIO/eos/pull/9033 & https://github.com/EOSIO/eos/pull/9209 & https://github.com/EOSIO/eos/pull/9317

### CODE REVIEW

File: tests/plugin_http_api_test.py [hunk 1]
```diff
@@ -0,0 +1,1344 @@
...
+        # abi_bin_to_json with valid parameter
+        valid_cmd = ("%s%s '{%s,%s,%s}'") % (default_cmd,
+                                             self.http_post_str,
+                                             "\"code\":\"eosio.token\"",
+                                             "\"action\":\"issue\"",
+                                             "\"args\":\"ee6fff5a5c02c55b6304000000000100a6823403ea3055000000572d3ccdcd0100000000007015d600000000a8ed32322a00000000007015d6000000005c95b1ca102700000000000004454f53000000000968656c6c6f206d616e00\"")
+        ret_json = Utils.runCmdReturnJson(valid_cmd)
+        self.assertEqual(ret_json["code"], 500)
+
+        # get_required_keys with empty p4arameter
```
Comment (ClaytonCalabrese) [inline tests/plugin_http_api_test.py:492]:
```suggestion
        # get_required_keys with empty parameter
```

File: unittests/plugin_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+BOOST_AUTO_TEST_SUITE(plugin_tests)
+
+BOOST_AUTO_TEST_CASE( make_trimmed_string_view ) try {
+   {  // empty string
+      const std::string empty_str;
+      BOOST_REQUIRE(empty_str.empty());
+      BOOST_REQUIRE(eosio::make_trimmed_string_view(empty_str).empty());
+   }
+   {  // all spacee string
```
Comment (ClaytonCalabrese) [inline unittests/plugin_tests.cpp:35]:
```suggestion
   {  // all space string
```

File: unittests/plugin_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+      BOOST_CHECK_EQUAL(eosio::make_trimmed_string_view(str_with_leading_space), str_no_pace_both_side);
+
+      const std::string str_with_with_trailing_space = str_no_pace_both_side + " ";
+      BOOST_REQUIRE(!str_with_with_trailing_space.empty());
+      BOOST_CHECK_EQUAL(eosio::make_trimmed_string_view(str_with_with_trailing_space), str_no_pace_both_side);
+
+      const std::string str_with_with_space_both_side = " " + str_no_pace_both_side + " ";
+      BOOST_REQUIRE(!str_with_with_space_both_side.empty());
+      BOOST_CHECK_EQUAL(eosio::make_trimmed_string_view(str_with_with_space_both_side), str_no_pace_both_side);
+   }
```
Comment (ClaytonCalabrese) [inline unittests/plugin_tests.cpp:75]:
```suggestion
      const std::string str_no_space_both_side  = "a b";
      BOOST_REQUIRE(!str_no_space_both_side.empty());
      BOOST_CHECK_EQUAL(eosio::make_trimmed_string_view(str_no_space_both_side), str_no_space_both_side);

      const std::string str_with_leading_space = " " + str_no_space_both_side;
      BOOST_REQUIRE(!str_with_leading_space.empty());
      BOOST_CHECK_EQUAL(eosio::make_trimmed_string_view(str_with_leading_space), str_no_space_both_side);

      const std::string str_with_with_trailing_space = str_no_space_both_side + " ";
      BOOST_REQUIRE(!str_with_with_trailing_space.empty());
      BOOST_CHECK_EQUAL(eosio::make_trimmed_string_view(str_with_with_trailing_space), str_no_space_both_side);

      const std::string str_with_with_space_both_side = " " + str_no_space_both_side + " ";
      BOOST_REQUIRE(!str_with_with_space_both_side.empty());
      BOOST_CHECK_EQUAL(eosio::make_trimmed_string_view(str_with_with_space_both_side), str_no_space_both_side);
   }
```

File: unittests/plugin_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+      const std::string empty_content_str_3 = "{\" \"}";
+      BOOST_REQUIRE(!empty_content_str_3.empty());
+      BOOST_REQUIRE(eosio::is_empty_content(empty_content_str_3) == false);
+   }
+   {  // invalid string
+      const std::string invalid_str_1 = "{  a";
+      BOOST_REQUIRE(!invalid_str_1.empty());
+      BOOST_REQUIRE(eosio::is_empty_content(invalid_str_1) == false);
+      const std::string invalid_str_2 = " a }";
+      BOOST_REQUIRE(! invalid_str_2.empty());
```
Comment (ClaytonCalabrese) [inline unittests/plugin_tests.cpp:131]:
```suggestion
      BOOST_REQUIRE(!invalid_str_2.empty());
```

**Review by ClaytonCalabrese — CHANGES_REQUESTED** [review]

File: plugins/producer_api_plugin/producer_api_plugin.cpp [hunk 5]
```diff
@@ -59,28 +58,27 @@ struct async_result_visitor : public fc::visitor<fc::variant> {
...
 
 #define INVOKE_R_V_ASYNC(api_handle, call_name)\
      api_handle.call_name(next);
 
 #define INVOKE_V_R(api_handle, call_name, in_param) \
-     api_handle.call_name(fc::json::from_string(body).as<in_param>()); \
-     eosio::detail::producer_api_plugin_response result{"ok"};
-
-#define INVOKE_V_R_R(api_handle, call_name, in_param0, in_param1) \
-     const auto& vs = fc::json::json::from_string(body).as<fc::variants>(); \
```
[the remark this answers is not in the document]
Comment (heifner) [reply plugins/producer_api_plugin/producer_api_plugin.cpp:79]: V = void, here it means it returns a void
R = param. So R_R means a method taking two params.

File: plugins/http_plugin/include/eosio/http_plugin/http_plugin.hpp [hunk 6]
```diff
@@ -157,6 +158,93 @@ namespace eosio {
...
+            }
+         }
+      } else {
+         return false;
+      }
+      return true;
+   }
+
+   enum class http_params_types {
+      no_params_required = 0,
```
[the remark this answers is not in the document]
Comment (heifner) [reply plugins/http_plugin/include/eosio/http_plugin/http_plugin.hpp]: renamed

**Review by ClaytonCalabrese — APPROVED** [review]

**Pull request merged**